### PR TITLE
Adds click attacks

### PR DESCRIPTION
**Pull request by Zumorica**
at long last, eh?

Requires a submodule update.

### CODE REVIEW

**Comment by SweptWasTaken** [comment]
You should take away things turning red when you hit them. There's already a hit effect.

**Comment by Zumorica** [comment]
> You should take away things turning red when you hit them. There's already a hit effect.

Nah. I personally like the effect, feels like you actually hit the thing in question.

File: Content.Server/GameObjects/Components/Weapon/Melee/MeleeWeaponComponent.cs [hunk 1]
```diff
@@ -17,79 +17,78 @@
...
 
         protected virtual bool OnHitEntities(IReadOnlyList<IEntity> entities, AttackEventArgs eventArgs)
         {
             return true;
         }
 
-        void IAttack.Attack(AttackEventArgs eventArgs)
+        bool IAttack.WideAttack(AttackEventArgs eventArgs)
         {
+            if (!eventArgs.WideAttack) return true;
```
Comment (DrSmugleaf) [inline Content.Server/GameObjects/Components/Weapon/Melee/MeleeWeaponComponent.cs:85]: This should return false to let other IAttack interfaces to run... Even though the return type isn't used anywhere yet. Just for consistency with the rest of the interaction interfaces.

File: Content.Server/GameObjects/Components/Weapon/Melee/MeleeWeaponComponent.cs [hunk 2]
```diff
@@ -17,79 +17,78 @@
...
+
             var curTime = IoCManager.Resolve<IGameTiming>().CurTime;
-            var span = curTime - _lastAttackTime;
-            if(span.TotalSeconds < _cooldownTime) {
-                return;
-            }
+
+            if(curTime < _cooldownEnd)
+                return true;
```
Comment (DrSmugleaf) [inline Content.Server/GameObjects/Components/Weapon/Melee/MeleeWeaponComponent.cs:90]: Same here.

File: Content.Server/GameObjects/Components/Weapon/Melee/MeleeWeaponComponent.cs [hunk 3]
```diff
@@ -114,26 +113,82 @@ void IAttack.Attack(AttackEventArgs eventArgs)
...
+        }
+
+        bool IAttack.ClickAttack(AttackEventArgs eventArgs)
+        {
+            if (eventArgs.WideAttack) return false;
+
+            var curTime = IoCManager.Resolve<IGameTiming>().CurTime;
+
+            if(curTime < _cooldownEnd || !eventArgs.Target.IsValid())
+                return true;
```
Comment (DrSmugleaf) [inline Content.Server/GameObjects/Components/Weapon/Melee/MeleeWeaponComponent.cs:148]: Same here.

File: Content.Shared/Interfaces/GameObjects/Components/Interaction/IAttack.cs [hunk 4]
```diff
@@ -9,18 +13,25 @@ namespace Content.Shared.Interfaces.GameObjects.Components
...
             ClickLocation = clickLocation;
+            WideAttack = wideAttack;
+            Target = target;
         }
 
         public IEntity User { get; }
         public GridCoordinates ClickLocation { get; }
+        public bool WideAttack { get; }
+        public EntityUid Target { get; }
+        public IEntity? TargetEntity => Target.IsValid() ? IoCManager.Resolve<IEntityManager>()?.GetEntity(Target) ?? null : null;
```
Comment (DrSmugleaf) [inline Content.Shared/Interfaces/GameObjects/Components/Interaction/IAttack.cs:35]: This should be changed to be an assignment in the constructor.

File: Content.Shared/Interfaces/GameObjects/Components/Interaction/IAttack.cs [hunk 5]
```diff
@@ -9,18 +13,25 @@ namespace Content.Shared.Interfaces.GameObjects.Components
     /// </summary>
     public interface IAttack
     {
-        void Attack(AttackEventArgs eventArgs);
+        // Redirects to ClickAttack by default.
```
Comment (DrSmugleaf) [inline Content.Shared/Interfaces/GameObjects/Components/Interaction/IAttack.cs:16]: What's more likely to happen, wide attacks or click attacks? Whichever it is should decide the redirect. I'm not sure personally, so I would say that this is fine unless you want it to be more complicated with default/virtual methods.

**Review by DrSmugleaf — CHANGES_REQUESTED** [review]

File: Content.Shared/Interfaces/GameObjects/Components/Interaction/IAttack.cs (hunk 5 above)
Comment (Zumorica) [reply Content.Shared/Interfaces/GameObjects/Components/Interaction/IAttack.cs:16]: Click attacks are more likely to happen imo

File: Content.Server/GameObjects/Components/Weapon/Melee/MeleeWeaponComponent.cs (hunk 2 above)
Comment (Zumorica) [reply Content.Server/GameObjects/Components/Weapon/Melee/MeleeWeaponComponent.cs:91]: this needs to return true so it doesn't default to attacking with the fist.

File: Content.Server/GameObjects/Components/Weapon/Melee/MeleeWeaponComponent.cs (hunk 3 above)
Comment (Zumorica) [reply Content.Server/GameObjects/Components/Weapon/Melee/MeleeWeaponComponent.cs:149]: this needs to return true so it doesn't default to attacking with the fist

**Pull request merged**